cloudscale_server: refactor to implement update

resmo opened the pull request:
##### SUMMARY
Refactored to implement the update functionality, it has been tested with exiting tests before adding the new tests covering the updating parts. Further the refactored module has `--diff` support implemented.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
cloudscale_server

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

resmo committed:
cloudscale_server: remove previous version of module

resmo committed:
cloudscale_server: add refactored module with update implemented

resmo committed:
cloudscale: move common code to module_utils

resmo committed:
tests: adjust existing tests to pass

resmo committed:
tests: cloudscale_server: extend with update tests

resmo committed:
fix sanity

resmo commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+            server_info = self._wait_for_state(('running', ))
+        return server_info
+
+    def _has_changed(self, server_info, data):
```
probably rename this method to `_get_keys_changed` to reflect the actually behaviour

resmo commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+        if not keys_changed:
+            return server_info
+
+        # We have changed keys but we ignore them on unless forced if server is running
```
the word `on` is too much

resmo committed:
cloudscale_server: minor adjustments

resmo commented on lib/ansible/module_utils/cloudscale.py:
```diff
@@ -44,7 +49,15 @@ def _get(self, api_call):
+            # Deepcopy: Duplicate the data object for iteration, because
+            # iterating an object and changing it at the same time is insecure
+            for k, v in deepcopy(data).items():
+                for k, v in data.items():
```
ehm ok, this does not look the way it should.

gaudenz replied:
You probably want to do something like `dict((k,v) for k,v in data.items() if v is not None)`. With this a new dict is generated and there is no deep-copying needed. But on the other hand it's unclear to me why you actually have to delete `None` values.

gaudenz commented on lib/ansible/module_utils/cloudscale.py:
```diff
@@ -44,7 +44,15 @@ def _get(self, api_call):
     def _post_or_patch(self, api_call, method, data):
         headers = self._auth_header.copy()
-        if data is not None:
+        if data:
```
This looks wrong or is at least a change in behavior. With the new clause you can no longer send an empty dict as data. Not sure if there is a call where this would make sense, but just wanted to note that there is a subtle difference here. I would prefer the `if data is not None:` variant.

resmo replied:
can't see neither why I made this reverting...

gaudenz commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+    ssh_keys: ssh-rsa XXXXXXXXXXX ansible@cloudscale
+    api_token: xxxxxx
+  register: server
+  until: server.ssh_fingerprints is defined
```
AFAIK this test is wrong. The previous version had `until: server.ssh_fingerprints`. The API always returns the ssh_fingerprints key, but until the fingerprints are available the value is `null`. So the new condition is always true even if the fingerprints are not yet available. If you are concerned that the key might not be available (this should never happen) you might want to write it as `until: server.ssh_fingerprints is defined and server.ssh_fingerprints`.

gaudenz commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+notes:
+  - If more than one server with the name given by the I(name) option exists, execution is aborted.
+  - Only the I(name) and I(flavor) are evaluated for the update.
+  - The option I(force=true) must be given to allow the reboot of existing running servers for applying the changes.
```
Maybe we should be more verbose about the interaction of I(name) and I(uuid) here. If I understand the intention correctly the following happens:
- If name and uuid are given and the name of the server with this uuid does not match the name parameter, the server is renamed.
- If multiple servers with the same name as the name parameter exist and no uuid is given, the execution is aborted.
- In all other cases either name or uuid can be used as keys to select an existing server.

Is this correct?

resmo replied:
this is correct

gaudenz commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+        if server_info.get('status') == "running":
+            if not self._module.params.get('force'):
+                self._module.warn("Changes won't be applied to running servers. "
+                                  "Use force=yes to allow the server '%s' to be stopped/started." % server_info['name'])
```
This also forces a server to be stopped for a rename. The cloudscale.ch API does not require this. Running servers can be renamed.
For the case where the flavor changes I'm not sure if I like the warning or if I'd prefer the module to fail. IMO modules should fail when they can't do a requested action. Or is there an Ansible policy that this should only be a warning?

gaudenz commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+                                  "Use force=yes to allow the server '%s' to be stopped/started." % server_info['name'])
+                return server_info
+            else:
+                self._module.warn("The server '%s' will be stopped/started to apply changes because force=true is provided" % server_info['name'])
```
I don't like this warning. If the force parameter is given, this is expected behavior. So no warning is needed. It just clutters the Ansible output. Logging something that is not shown by default seems enough if needed.

gaudenz commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+
+        self._info = self._init_server_container()
+
+        uuid = self._module.params.get('uuid') or self._info.get('uuid')
```
As you already have the same 'or' clause in `_init_server_container` you can just rely on `self._info.get('uuid')` here.

gaudenz commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,500 @@
+        # Keys to transform in get_returns()
+        self._transforms = {
+            'status': dict(to_key='state'),
+        }
```
I would probably make this a module level constant like `ALLOWED_STATES`.
I also noticed that you don't transform the output for the diff. I don't think it matters much but for consistency this should ideally also be transformed.
Maybe the best way to handle this would be to transform the values already in `def _get_server_info`. This way the module internal name would always be "state" and there would be no confusion between "state" and "status". That's also how this was handled before your change.

resmo commented:
waiting for https://github.com/ansible/ansible/pull/52806 to fix the false negative sanity failure

resmo commented:
ready_for_review

gaudenz commented on lib/ansible/module_utils/cloudscale.py:
```diff
@@ -88,3 +93,13 @@ def _delete(self, api_call):
+            for k, v in resource.items():
+                if k in self._transforms:
+                    new_key = self._transforms[k]['to_key']
+                    self._result[new_key] = v
```
As already noted elsewhere I would prefer the transformations to be done as early as possible and not as late as possible just before returning the data because it makes the code easier to understand.

gaudenz commented on test/integration/targets/cloudscale_server/tasks/main.yml:
```diff
@@ -39,6 +39,92 @@
+  assert:
+    that:
+      - server is changed
+      - server.state == 'running'
```
This should also check `server.flavor == cloudscale_test_flavor` to ensure the flavor is not changed in check mode.

This also applies to the other tests above if we keep the current behavior of the force parameter. But see my other comment on that.

gaudenz commented on test/integration/targets/cloudscale_server/tasks/main.yml:
```diff
@@ -39,6 +39,92 @@
+  assert:
+    that:
+      - server is changed
+      - server.state == 'running'
```
Same here this should also check for the actual change requested: `server.flavor == cloudscale_test_flavor_2`. The same applies for all tests further down.

gaudenz commented on test/integration/targets/cloudscale_server/tasks/main.yml:
```diff
@@ -127,6 +213,49 @@
+      - server is not changed
+      - server.state == 'stopped'
+
 - name: Test server running in check mode
```
An additional test to change the server name would be nice. I don't think we need the full set of force yes/no check mode and idempotence tests here. IMO just one test with check mode yes and no would be enough. if you want to be fancy, add an idempotence test.

gaudenz reviewed (changes requested):
Thanks for your work! It's nice to see this module being improved. Please also have a look at my suggestion regarding the code you already fixed since I started my review.

resmo committed:
check for None

resmo committed:
example check ssh_fingerprints for defined and falsey

resmo committed:
reduce verbosity by removing a warning

resmo committed:
transform status to state early

resmo committed:
don't force stop for name changes, extends tests

resmo committed:
minor cleanup

resmo commented:
I addressed the points of the review, clarifying docs is open yet. However, the update method has a lot of duplicate logic and looks over complicated, I am refactoring to simplify it.

resmo committed:
simplify update logic

resmo committed:
clarify docs

resmo commented:
ready_for_review

resmo committed:
simplify update_param even more

gaudenz commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -252,80 +291,153 @@ def _transform_state(server):
+            # Either the server is stopped or change is forced
+            self._result['changed'] = True
+            if not self._module.check_mode:
+                # Remember the state of the server before we ensure it is stopped
```
With the changed code this comment is now wrong. Please move to the place where the server state is actually saved.

gaudenz reviewed (changes requested):
Thanks for all the changes. There is only one issue from my review left that I would like to discuss:
> For the case where the flavor changes I'm not sure if I like the warning or if I'd prefer the module to fail. IMO modules should fail when they can't do a requested action. Or is there an Ansible policy that this should only be a warning?

resmo committed:
get rid of mislieading comment

resmo commented:
I don't like the idea to fail. It is not an error, it is just not clear if the user does agree to take the server offline to change it. Hence the force param and a message to the user. 

(Hypothetical, if we would fail, shouldn't we then also fail on other params which the api doesn't allow to update?)

Last but not least there are other cloud modules using the same behavior.

resmo commented:
ready_for_review

gaudenz commented:
Thanks for the last small change and the explanation.

shipit
